java.hints/src/org/netbeans/modules/java/hints/jdk/ConvertToVarHint.java

```diff
@@ -160,4 +167,36 @@ private static boolean isDiagnosticCodeTobeSkipped(CompilationInfo info) {
         List<Diagnostic> diagnosticsList = info.getDiagnostics();
         return diagnosticsList.stream().anyMatch((d) -> (SKIPPED_ERROR_CODES.contains(d.getCode())));
     }
+    
+    private static boolean isValidVarType(HintContext ctx) {
+        TreePath treePath = ctx.getPath();
+        TreePath initTreePath = ctx.getVariables().get("$init");  //NOI18N
+
+        TypeMirror variableTypeMirror = ctx.getInfo().getTrees().getElement(treePath).asType();
+
+        if (initTreePath.getLeaf().getKind() == Tree.Kind.NEW_CLASS) {
+            NewClassTree nct = (NewClassTree) (initTreePath.getLeaf());
+            //anonymous class type
+            if (nct.getClassBody() != null) {
+                return false;
+            }
+        } else if (initTreePath.getLeaf().getKind() == Tree.Kind.NEW_ARRAY) {
+            NewArrayTree nat = (NewArrayTree) ((VariableTree) treePath.getLeaf()).getInitializer();
+            //array initializer expr type
+            if (nat.getType() == null) {
+                return false;
+            }
+        } else if (initTreePath.getLeaf().getKind() == Tree.Kind.LAMBDA_EXPRESSION) {
+            return false;
+        }
+        // variable initializer type should be same as variable type.
+        TypeMirror initTypeMirror = ctx.getInfo().getTrees().getTypeMirror(initTreePath);
+        
+        //TODO: add support for CapturedType TypeParams: testcase testCapturedTypeTypeParamsAssignToVar3()
+        if (!(initTypeMirror instanceof CapturedType) && (!ctx.getInfo().getTypes().isSameType(variableTypeMirror, initTypeMirror))) {
```

Comment:
Please no access to javac internal types (com.sun.tools.**) in hint implementations. Maybe initTypeMirror = org.netbeans.modules.java.hints.errors.Utilities.resolveCapturedType(initTypeMirror) + just he isSameType check would work? (Possibly check that initTypeMirror is not an erroneous type before, not sure.)

Reply:
Done, Thanks